File: tactics2d/math/geometry/circle.py
##! python3
# Copyright (C) 2024, Tactics2D Authors. Released under the GNU GPLv3.
# @File: circle.py
# @Description: This file implements some frequently operations on circle.
# @Author: Tactics2D Team
# @Version: 1.0.0

from enum import Enum
from typing import Tuple, Union

import numpy as np
from cpp_function import Circle as cpp_circle


class Circle:
    """This class implement some frequently operations on circle.

    !!! note "TODO"
        To improve the performance, we will rewrite the methods in C++ in the future.
    """

    class ConstructBy(Enum):
        """The method to derive a circle.

        Attributes:
            ThreePoints (int): Derive a circle by three points.
            TangentVector (int): Derive a circle by a tangent point, a heading and a radius.
        """

        ThreePoints = 1
        TangentVector = 2

    @staticmethod
    def get_circle_by_three_points(
        pt1: Union[list, np.ndarray], pt2: Union[list, np.ndarray], pt3: Union[list, np.ndarray]
    ) -> Tuple[np.ndarray, float]:
        """This function gets a circle by three points.

        Args:
            pt1 (Union[list, np.ndarray]): The first point. The shape is (2,).
            pt2 (Union[list, np.ndarray]): The second point. The shape is (2,).
            pt3 (Union[list, np.ndarray]): The third point. The shape is (2,).

        Returns:
            center (np.ndarray): The center of the circle. The shape is (2,).
            radius (float): The radius of the circle.
        """
        ## Verify input
        assert len(pt1) == 2
        assert len(pt2) == 2
        assert len(pt3) == 2

        ## Derive the circle
        d = -np.linalg.det(
            [
                [pt1[0] ** 2 + pt1[1] ** 2, pt1[1], 1],
                [pt2[0] ** 2 + pt2[1] ** 2, pt2[1], 1],
                [pt3[0] ** 2 + pt3[1] ** 2, pt3[1], 1],
            ]
        )
        e = np.linalg.det(
            [
                [pt1[0] ** 2 + pt1[1] ** 2, pt1[0], 1],
                [pt2[0] ** 2 + pt2[1] ** 2, pt2[0], 1],
                [pt3[0] ** 2 + pt3[1] ** 2, pt3[0], 1],
            ]
        )
        det = np.linalg.det([[pt1[0], pt1[1], 1], [pt2[0], pt2[1], 1], [pt3[0], pt3[1], 1]])

        D = d / det
        E = e / det

        center = [-D / 2, -E / 2]
        radius = np.linalg.norm(pt1 - center)

        return center, radius

    @staticmethod
    def get_circle_by_tangent_vector(
        tangent_point: Union[list, np.ndarray], heading: float, radius: float, side: str
    ) -> Tuple[np.ndarray, float]:
        """This function gets a circle by a tangent point, a heading and a radius.

        Args:
            tangent_point (np.ndarray): The tangent point on the circle. The shape is (2,).
            heading (float): The heading of the tangent point. The unit is radian.
            radius (float): The radius of the circle.
            side (int): The location of circle center relative to the tangent point. "L" represents left. "R" represents right.

        Returns:
            center (np.ndarray): The center of the circle. The shape is (2,).
            radius (float): The radius of the circle.
        """
        if side == "R":
            vec = np.array([np.cos(heading - np.pi / 2), np.sin(heading - np.pi / 2)]) * radius
        elif side == "L":
            vec = np.array([np.cos(heading + np.pi / 2), np.sin(heading + np.pi / 2)]) * radius
        return tangent_point + vec, radius

    @staticmethod
    def get_circle(method: ConstructBy, *args: tuple):
        """This function gets a circle by different given conditions.

        Args:
            method (ConstructBy): The method to derive a circle. The available choices are Circle.ConstructBy.ThreePoints) and Circle.ConstructBy.TangentVector).
            *args (tuple): The arguments of the method.

        Returns:
            center (np.ndarray): The center of the circle. The shape is (2,).
            radius (float): The radius of the circle.

        Raises:
            NotImplementedError: The input method id is not an available choice.
        """
        if method == Circle.ConstructBy.ThreePoints:
            # return Circle.get_circle_by_three_points(*args)
            center, radius = cpp_circle.get_circle_by_three_points(*args)
            return np.array(center), radius
        elif method == Circle.ConstructBy.TangentVector:
            # return Circle.get_circle_by_tangent_vector(*args)
            center, radius = cpp_circle.get_circle_by_tangent_vector(*args)
            return np.array(center), radius

        raise NotImplementedError

    @staticmethod
    def get_arc(
        center_point: np.ndarray,
        radius: float,
        delta_angle: float,
        start_angle: float,
        clockwise: bool = True,
        step_size: float = 0.1,
    ) -> np.ndarray:
        """This function gets the points on an arc curve line.

        Args:
            center_point (np.ndarray): The center of the arc. The shape is (2,).
            radius (float): The radius of the arc.
            delta_angle (float): The angle of the arc. This values is expected to be positive. The unit is radian.
            start_angle (float): The start angle of the arc. The unit is radian.
            clockwise (bool): The direction of the arc. True represents clockwise. False represents counterclockwise.
            step_size (float): The step size of the arc. The unit is radian.

        Returns:
            arc_points(np.ndarray): The points on the arc. The shape is (int(radius * delta / step_size), 2).
        """
        if clockwise:
            angles = np.array(
                np.arange(start_angle, (start_angle - delta_angle), -step_size / radius)
            )
        else:
            angles = np.array(
                np.arange(start_angle, (start_angle + delta_angle), step_size / radius)
            )

        arc_points = center_point + np.array([np.cos(angles), np.sin(angles)]).T * radius

        # TODO:Frequent conversions to numpy arrays can slow down the runtime.
        # arc_points = np.array(cpp_circle.get_arc(center_point, radius, delta_angle, start_angle, clockwise, step_size))
        return arc_points

File: tactics2d/math/cpp_function/src/circle.cpp
// circle.cpp
#include "circle.h"


std::pair<std::array<double, 2>, double> Circle::get_circle_by_three_points(
    const std::array<double, 2>& pt1,
    const std::array<double, 2>& pt2,
    const std::array<double, 2>& pt3
) {
    /*This function gets a circle by three points.

    Args:
        pt1 (std::array<double, 2>): The first point. The shape is (2,).
        pt2 (std::array<double, 2>): The second point. The shape is (2,).
        pt3 (std::array<double, 2>): The third point. The shape is (2,).

    Returns:
        center (std::array<double, 2>): The center of the circle. The shape is (2,).
        radius (double): The radius of the circle.
   */

    double a = pt1[0] - pt2[0];
	double b = pt1[1] - pt2[1];
	double c = pt1[0] - pt3[0];
	double d = pt1[1] - pt3[1];
	double e = ((pt1[0] * pt1[0] - pt2[0] * pt2[0])-(pt2[1] * pt2[1] - pt1[1] * pt1[1])) / 2;
	double f = ((pt1[0] * pt1[0] - pt3[0] * pt3[0])-(pt3[1] * pt3[1] - pt1[1] * pt1[1])) / 2;
	

    double denom = a * d - b * c;
    if (std::abs(denom) < 1e-10) {
        throw std::runtime_error("Points are collinear");
    }


    double x = (e * d - b * f) / denom;
    double y = (a * f - e * c) / denom;

    std::array<double, 2> center = {x, y};

    double temp1 = pt1[0] - center[0];
    double temp2 = pt1[1] - center[1];
    double radius = std::sqrt(temp1 * temp1 + temp2 * temp2);

    return {center, radius};
}

std::pair<std::array<double, 2>, double> Circle::get_circle_by_tangent_vector(
    const std::array<double, 2>& tangent_point,
    double heading,
    double radius,
    const std::string& side
) {
    /*This function gets a circle by a tangent point, a heading and a radius.

    Args:
        tangent_point (std::array<double, 2>): The tangent point on the circle. The shape is (2,).
        heading (double): The heading of the tangent point. The unit is radian.
        radius (double): The radius of the circle.
        side (int): The location of circle center relative to the tangent point. "L" represents left. "R" represents right.

    Returns:
        center (std::array<double, 2>): The center of the circle. The shape is (2,).
        radius (double): The radius of the circle.
    */

    std::array<double, 2> vec;
    if (side == "R") {
        vec = {std::cos(heading - M_PI / 2) * radius, std::sin(heading - M_PI / 2) * radius};
    } else if (side == "L") {
        vec = {std::cos(heading + M_PI / 2) * radius, std::sin(heading + M_PI / 2) * radius};
    } else {
        throw std::invalid_argument("Invalid side. Use 'L' or 'R'.");
    }

    return { {tangent_point[0] + vec[0], tangent_point[1] + vec[1]}, radius };
}

std::pair<std::array<double, 2>, double> Circle::get_circle(ConstructBy method, py::args args) {
    /*This function gets a circle by different given conditions.

    Args:
        method (ConstructBy): The method to derive a circle. The available choices are Circle.ConstructBy.ThreePoints) and Circle.ConstructBy.TangentVector).
        *args (py::args): The arguments of the method.

    Returns:
        center (std::array<double, 2>): The center of the circle. The shape is (2,).
        radius (double: The radius of the circle.
    */

    if (method == ConstructBy::ThreePoints) {
        return get_circle_by_three_points(
            args[0].cast<std::array<double, 2>>(),
            args[1].cast<std::array<double, 2>>(),
            args[2].cast<std::array<double, 2>>()
        );
    } else if (method == ConstructBy::TangentVector) {
        if (args.size() != 4) {
            throw std::invalid_argument("TangentVector method requires 4 arguments");
        }
        return get_circle_by_tangent_vector(
            args[0].cast<std::array<double, 2>>(),
            args[1].cast<double>(),
            args[2].cast<double>(),
            args[3].cast<std::string>()
        );
    } else {
        throw std::invalid_argument("Invalid method");
    }    
}


std::vector<std::array<double, 2>> Circle::get_arc(
    const std::array<double, 2>& center_point,
    double radius,
    double delta_angle,
    double start_angle,
    bool clockwise,
    double step_size
) {
    /*This function gets the points on an arc curve line.

    Args:
        center_point (std::array<double, 2>): The center of the arc. The shape is (2,).
        radius (double): The radius of the arc.
        delta_angle (double): The angle of the arc. This values is expected to be positive. The unit is radian.
        start_angle (double): The start angle of the arc. The unit is radian.
        clockwise (bool): The direction of the arc. True represents clockwise. False represents counterclockwise.
        step_size (double): The step size of the arc. The unit is meter.

    Returns:
        arc_points(std::array<double, 2>): The points on the arc. The shape is (int(radius * delta / step_size), 2).
    */

    double angle_step = step_size / radius;
    int num_steps = static_cast<int>(std::ceil(delta_angle / angle_step));

    std::vector<std::array<double, 2>> arc_points_vec(num_steps);

    if (clockwise) {
        angle_step = -angle_step;
    } else {
        angle_step = angle_step;
    }
    double angle = start_angle - angle_step;
    for (int i = 0; i < num_steps; ++i) {
        
        angle += angle_step;

        arc_points_vec[i][0] = center_point[0] + radius * std::cos(angle);
        arc_points_vec[i][1] = center_point[1] + radius * std::sin(angle);
    }

    return arc_points_vec;
}

